### ISSUE TITLE
Implement "group0_voter" configuration variable

### ISSUE THREAD

**Issue by kostja** [scylladb/scylladb#22071]
To manually manage complex topology, it would be handy to be able to disable voter rights for specific nodes.
This would allow, e.g. to create "biased" clusters in 2-dc setup, where one DC has fewer votesrs, so the cluster can survive split brain by failing over to the primary DC.

This is a common deployment pattern.

The variable is relevant when a node joins the cluster. If a node restarts and its setting has changed, it could try to change its voting disposition, but this is not strictly necessary.

Alternately, we can disable voting for all but zero-token nodes, since group0 leadership is a heavy responsibility.

**Comment by kostja** [comment]
@kbr-scylla  @emdroid concluded we need this while working on my scylla summit presentaiton; doesn't seem to be hard to do, wdyt?

**Comment by emdroid** [comment]
Right, I assume it will be a rather smaller task. The simplest might be to make it a startup parameter for the node (currently the node voting is requested by the node itself). Although it might have some interference with the #21901 that I'm currently working on, where the voting status will be handled by the topology coordinator eventually (plus even without that, I think this task would also require some improved handling in the topology coordinator code that alters the voters status, for example during the node removal and rollback).

Within #21901 this might complicate it a bit, as the group0_voter configuration will need to be taken into account when managing the voters on the coordinator level (although probably solvable by simply ignoring the mode when pushing the nodes into the voter manager).

**Comment by kostja** [comment]
* Emil Maskovsky ***@***.***> [24/12/29 22:22]:
> Right, I assume it will be a rather smaller task. The simplest might be to make it a startup parameter for the node (currently the node voting is requested by the node itself). Although it might have some interference with the #21901 that I'm currently working on, where the voting status will be handled by the topology coordinator eventually (plus even without that, I think this task would also require some improved handling in the topology coordinator code that alters the voters status, for example during the node removal and rollback).
> 
> Within #21901 this might complicate it a bit, as the
> group0_voter configuration will need to be taken into account
> when managing the voters on the coordinator level (although
> probably solvable by simply ignoring the mode when pushing the
> nodes into the voter manager).

When the node joins the cluster it could pass it on to group0
state whether or not it wants to be part of electorate or not.


-- 
Konstantin Osipov

**Comment by kbr-scylla** [comment]
I'd like to avoid configuration variables when there's not sufficient motivation for them. And since we're already doing https://github.com/scylladb/scylladb/issues/18793 already, I don't see what value this new thing adds? https://github.com/scylladb/scylladb/issues/18793 should choose asymmetric voter set in symmetric 2 DC setup.

**Comment by emdroid** [comment]
Right, in fact it looks like the limited voters feature should actually address this. Only it could be tricky to tell which of the 2 DCs has more voters, but I think this could be made deterministic by using more racks for the DC where we want more voters (as a different rack means a "bigger distance", which will mean that the DC with more racks should get more voters at the end).

**Comment by kostja** [comment]
1) In large clusters, I want only dedicated zero-token nodes to be voters, since group0 leader role becomes heavier and heavier. 
2) in a 2-dc environment, I want one DC to be primary, another secondary. I.e. I want to survive in case of loss of a secondary DC.

**Comment by gleb-cloudius** [comment]
> 1. In large clusters, I want only dedicated zero-token nodes to be voters, since group0 leader role becomes heavier and heavier.

The algorithm to chose voters can prioritize zero toke nodes.

> 
>     2. in a 2-dc environment, I want one DC to be primary, another secondary. I.e. I want to survive in case of loss of a secondary DC.

This is a global setting, not per node one.

**Comment by kostja** [comment]
> > 1. In large clusters, I want only dedicated zero-token nodes to be voters, since group0 leader role becomes heavier and heavier.
> 
> The algorithm to chose voters can prioritize zero toke nodes.

This actually seems to make sense in all cases @emdroid 


> > 2. in a 2-dc environment, I want one DC to be primary, another secondary. I.e. I want to survive in case of loss of a secondary DC.
> > ```
> 
> This is a global setting, not per node one.

We don't have an entity  for DC in schema/metadata. So there is no place to specify this setting. Besides, that would imply complication of the voter selection, which would have to be DC aware, whereas if one can disable specific nodes, that opens an array of strategies to DBA, all satisfied with the same voter selection algorithm.

**Comment by gleb-cloudius** [comment]
> > > 2. in a 2-dc environment, I want one DC to be primary, another secondary. I.e. I want to survive in case of loss of a secondary DC.
> > 
> > 
> > This is a global setting, not per node one.
> 
> We don't have an entity for DC in schema/metadata. So there is no place to specify this setting. Besides, that would imply complication of the voter selection, which would have to be DC aware, whereas if one can disable specific nodes, that opens an array of strategies to DBA, all satisfied with the same voter selection algorithm.

Voter selection should be DC/rack aware anyway since we want to spread voters between them. But I do not see how disabling individual nodes will give you desired effect since it will only mean those nodes will not be considered by the algorithm, but the algorithm will still try to have equal amount of voters in each DC. You need to give an input to the algorithm that one DC is preferable. We can extend snitch with the configuration (may be).

**Comment by emdroid** [comment]
> > > 1. In large clusters, I want only dedicated zero-token nodes to be voters, since group0 leader role becomes heavier and heavier.
> > 
> > 
> > The algorithm to chose voters can prioritize zero toke nodes.
> 
> This actually seems to make sense in all cases [@emdroid](https://github.com/emdroid)

We had some discussion about this in the team before, as I also thought that it might make sense to prefer the zero-token nodes when choosing the voters. But the thought was it isn't necessary to handle the zero-token nodes in any different way than normal modes - see also the discussion in https://scylladb-corporate.slack.com/archives/G01F9RJQ95E/p1733412922498609?thread_ts=1733305144.737619&cid=G01F9RJQ95E - cc @patjed41

**Comment by emdroid** [comment]
The task to prioritize the zero-token nodes: #23267
